File: WolfDen/Lobbies/LobbyBehaviour/SlotColor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using WolfDen.Models;

namespace WolfDen.Lobbies.LobbyBehaviour
{
    public class SlotColor
    {
        public string Class { get { return GetType().Name; } }
        public Guid Id { get; } = Guid.NewGuid();
        public PlayerColor Color { get; set; }
        public List<PlayerColor> Colors { get; set; }
        public List<PlayerColor> AvailableColors { get; set; }
        public Slot Slot { get; set; }

        public SlotColor(Slot slot, List<PlayerColor> colors, List<PlayerColor> availableColors, PlayerColor playerColor)
        {
            Colors = colors;
            AvailableColors = availableColors;
            Color = playerColor;
            Slot = slot;
        }

        public bool ChangeColor(string ConnectionId, int colorId)
        {
            if (Slot?.LobbyPlayer?.ConnectionId != ConnectionId)
                return false;

            var newColor = AvailableColors.Where(c => c.ID == colorId).FirstOrDefault();
            if (newColor != null)
            {
                AvailableColors.Add(Color);
                AvailableColors.Remove(newColor);
                Color = newColor;
                return true;
            }
            return false;
        }
    }
}

File: WolfDen/wwwroot/ts/Lobby/LobbyBehaviour/SlotColor.ts
namespace TurnBasedBoardGame {
    export class SlotColor {
        Slot: Slot;
        _color: PlayerColor;
        static Colors: PlayerColor[];
        static AvailableColors: PlayerColor[];

        set AvailableColors(value: PlayerColor[]) {
            SlotColor.AvailableColors = value;
        }

        get AvailableColors(): PlayerColor[] {
            return SlotColor.AvailableColors;
        }

        get Color(): PlayerColor {
            return this._color;
        }

        set Color(value: PlayerColor) {
            this._color = value;
            $("#slot" + this.Slot?.Index + " .slotColor").css("background-color", value.Name);
            if (Lobby.Instance.GetCurrentSlot() == this.Slot)
                $("#color").css("background-color", value.Name);
        }

        constructor() {
            let self = this;
            $(".grid-color-option").click(function () {
                if (self.Slot.IsCurrentPlayer)
                    self.SelectColor($(this).attr("color"));
            });
        }

        Initialize() {
            // This is needed to set the html because this.Slot is not available when color is first set.
            this.Color = this.Color;
        }

        SelectColor(colorId: string) {
            $('#colorModal').modal('hide');
            var lobby = Lobby.Instance;
            lobby.Connection.invoke("ChangeColor", lobby.Id, this.Slot.Id, parseInt(colorId));
        }        

        ChangeColor(colorId: string) {
            var indexOfNewColor = SlotColor.AvailableColors.findIndex(c => c.ID == colorId);
            var newColor = SlotColor.AvailableColors[indexOfNewColor];
            SlotColor.AvailableColors[indexOfNewColor] = this.Color;
            this.Color = newColor;
        }
    }
}